### ISSUE TITLE
Does flannel with vxlan require manual entry for docker container IPs in ARP?

### ISSUE THREAD

**Issue by sanjana-bhat** [coreos/flannel#203]
Hi,

I'm running flanneld on 2 hosts which use vxlan device. When I try to ping from hostA docker container to the hostB docker container, it fails. The reason being, missing ARP entry for the hostB container IP or even hostB docker0 interface. I manually add the ARP entry for hostB container IP with the device address of hostB flannel.1 in hostA and the ping worked. Does flannel expect us to do this or should flannel daemon be taking care of ARP entries of docker containers?

arp -n
Address                  HWtype  HWaddress           Flags Mask            Iface
192.168.160.0            ether   42:63:3d:36:49:04   CM                    flannel.1
192.168.208.0            ether   ea:7e:ef:56:77:80   CM                    flannel.1
192.168.208.2            ether   02:42:c0:a8:d0:02   C                     docker0

As you can see from the above, there is only an entry for flannel.1 interface of hostB but not it's container IP which is 192.168.160.2 or it's docker0 interface which is 192.168.160.1

I build flannel locally and here is how I run it:
sudo bin/flanneld -etcd-endpoints="http://localhost:4001" -etcd-prefix="/coreos.com/network" -subnet-file="/run/flannel/subnet.env" -v=2 >flannel.log 2>&1 &

Please let me know if I'm missing something here. 

Thanks!

**Comment by eyakubovich** [comment]
@sanjana-bhat Are you building from git HEAD or from latest tag (v0.4.1). We just merged a change to use static/permanent entries instead of dynamic ones (https://github.com/coreos/flannel/pull/191) so it maybe related to that. If you're using HEAD, please try the tagged version and let me know if it helps.

You should not need to manually add ARP entries, it should just work. However, on hostA, hostB's container IP/ARP pairs should not exist. Only the hostB's `flannel.1` IP/pair should exist. The reason is that while we use VXLAN, it actually works at L3. The packets from the container are routed to flannel.1 VXLAN interface (via host's routing table) and then VXLAN is used to get the packet to remote flannel.1 interface. From there on, it's routed again to the docker0 bridge and into the container.

**Comment by eyakubovich** [comment]
@sanjana-bhat After looking into this, it is a problem with the commit I mentioned. v0.4.1 should work fine.

/cc @MohdAhmad

**Comment by sanjana-bhat** [comment]
@eyakubovich, thanks for a quick turnaround. After you pointed me to the tagged versions of flannel, I tried testing with both v0.4.1 and v0.4.0. The host from where I ping seems to crash soon after it receives 1 ICMP reply on flannel.1 device. 

Here is some of the information that can be useful
netstat -rn
Kernel IP routing table
Destination     Gateway         Genmask         Flags   MSS Window  irtt Iface
0.0.0.0         10.88.216.1     0.0.0.0         UG        0 0          0 eth0
10.88.216.0     0.0.0.0         255.255.252.0   U         0 0          0 eth0
169.254.0.0     0.0.0.0         255.255.0.0     U         0 0          0 eth0
192.0.0.0       0.0.0.0         255.0.0.0       U         0 0          0 flannel.1
192.168.208.0   0.0.0.0         255.255.240.0   U         0 0          0 docker0

bridge fdb
33:33:00:00:00:01 dev eth0 self permanent
01:00:5e:00:00:01 dev eth0 self permanent
33:33:ff:a6:d4:5a dev eth0 self permanent
b2:b3:c4:3d:89:c3 dev veth603752d vlan 0 permanent
02:42:c0:a8:d0:02 dev veth603752d vlan 0 
56:84:7a:fe:97:99 dev docker0 vlan 0 permanent
33:33:00:00:00:01 dev veth603752d self permanent
01:00:5e:00:00:01 dev veth603752d self permanent
33:33:ff:3d:89:c3 dev veth603752d self permanent

ping 192.168.176.2
PING 192.168.176.2 (192.168.176.2) 56(84) bytes of data.
64 bytes from 192.168.176.2: icmp_seq=2 ttl=62 time=38.6 ms

sudo tcpdump -i flannel.1
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on flannel.1, link-type EN10MB (Ethernet), capture size 65535 bytes
00:17:19.846838 IP6 hostA > ff02::2: ICMP6, router solicitation, length 16
00:20:28.241154 IP hostA > 192.168.176.2: ICMP echo request, id 3337, seq 1, length 64
00:20:29.239824 IP hostA > 192.168.176.2: ICMP echo request, id 3337, seq 2, length 64
00:20:29.276156 IP 192.168.176.2 > hostA: ICMP echo reply, id 3337, seq 2, length 64
00:20:30.241376 IP hostA > 192.168.176.2: ICMP echo request, id 3337, seq 3, length 64

dmesg tail
[    6.556765] systemd-journald[373]: Received request to flush runtime journal from PID 1
[    6.817361] type=1305 audit(1433720697.103:3): audit_pid=409 old=0 auid=4294967295 ses=4294967295 res=1
[   19.876463] EXT4-fs (vda): resizing filesystem from 262144 to 6553600 blocks
[   20.081281] EXT4-fs (vda): resized filesystem to 6553600
[   22.254248] serial8250: too much work for irq4
[   22.356039] serial8250: too much work for irq4
[   22.485038] serial8250: too much work for irq4
[   22.572043] serial8250: too much work for irq4
[   22.717039] serial8250: too much work for irq4
[   23.881928] serial8250: too much work for irq4
[   23.991117] serial8250: too much work for irq4
[   24.146457] serial8250: too much work for irq4
[   24.208504] serial8250: too much work for irq4
[   24.255481] serial8250: too much work for irq4
[  268.151009] [sched_delayed] sched: RT throttling activated
[  295.879342] ylock: module verification failed: signature and/or required key missing - tainting kernel
[  295.883888] ylock 1.102.0 loaded
[  307.948301] loop: module loaded
[  307.979247] bio: create slab <bio-1> at 1
[  308.006596] bio: create slab <bio-2> at 2
[  317.691314] Bridge firewalling registered
[  317.694090] IPv6: ADDRCONF(NETDEV_UP): docker0: link is not ready
[  317.761186] ip_tables: (C) 2000-2006 Netfilter Core Team
[  317.802295] nf_conntrack version 0.5.0 (16384 buckets, 65536 max)
[ 1006.084718] bio: create slab <bio-0> at 0
[ 1006.116622] IPv6: ADDRCONF(NETDEV_UP): docker0: link is not ready
[ 1030.944047] bio: create slab <bio-0> at 0
[ 1090.182251] IPv6: ADDRCONF(NETDEV_UP): docker0: link is not ready
[ 1128.887673] bio: create slab <bio-2> at 2
[ 1128.920954] EXT4-fs (dm-1): mounted filesystem with ordered data mode. Opts: discard
[ 1245.018069] bio: create slab <bio-2> at 2
[ 1245.166210] EXT4-fs (dm-1): mounted filesystem with ordered data mode. Opts: discard
[ 1245.382106] bio: create slab <bio-2> at 2
[ 1245.405185] EXT4-fs (dm-1): mounted filesystem with ordered data mode. Opts: discard
[ 1245.574347] bio: create slab <bio-2> at 2
[ 1245.594367] EXT4-fs (dm-1): mounted filesystem with ordered data mode. Opts: discard
[ 1255.429077] bio: create slab <bio-2> at 2
[ 1255.445620] EXT4-fs (dm-1): mounted filesystem with ordered data mode. Opts: discard
[ 1255.612680] bio: create slab <bio-2> at 2
[ 1255.636736] EXT4-fs (dm-1): mounted filesystem with ordered data mode. Opts: discard
[ 1255.908064] bio: create slab <bio-2> at 2
[ 1255.946280] EXT4-fs (dm-1): mounted filesystem with ordered data mode. Opts: discard
[ 1287.771814] bio: create slab <bio-2> at 2
[ 1287.797438] EXT4-fs (dm-1): mounted filesystem with ordered data mode. Opts: discard
[ 1288.164541] bio: create slab <bio-2> at 2
[ 1288.185703] EXT4-fs (dm-1): mounted filesystem with ordered data mode. Opts: discard
[ 1288.721778] bio: create slab <bio-2> at 2
[ 1288.751113] EXT4-fs (dm-1): mounted filesystem with ordered data mode. Opts: discard
[ 1289.575953] bio: create slab <bio-2> at 2
[ 1289.603117] EXT4-fs (dm-1): mounted filesystem with ordered data mode. Opts: discard
[ 1290.187508] bio: create slab <bio-2> at 2
[ 1290.212144] EXT4-fs (dm-1): mounted filesystem with ordered data mode. Opts: discard
[ 1290.694540] bio: create slab <bio-2> at 2
[ 1290.721100] EXT4-fs (dm-1): mounted filesystem with ordered data mode. Opts: discard
[ 1302.764889] bio: create slab <bio-2> at 2
[ 1302.791430] EXT4-fs (dm-1): mounted filesystem with ordered data mode. Opts: discard
[ 1306.340124] bio: create slab <bio-2> at 2
[ 1306.363347] EXT4-fs (dm-1): mounted filesystem with ordered data mode. Opts: discard
[ 1350.028090] device flannel.1 entered promiscuous mode
[ 1385.154255] bio: create slab <bio-2> at 2
[ 1385.222796] EXT4-fs (dm-1): mounted filesystem with ordered data mode. Opts: discard
[ 1385.806667] bio: create slab <bio-2> at 2
[ 1385.830508] EXT4-fs (dm-1): mounted filesystem with ordered data mode. Opts: discard
[ 1386.371750] bio: create slab <bio-2> at 2
[ 1386.398672] EXT4-fs (dm-1): mounted filesystem with ordered data mode. Opts: discard
[ 1386.996285] bio: create slab <bio-2> at 2
[ 1387.019808] EXT4-fs (dm-1): mounted filesystem with ordered data mode. Opts: discard
[ 1388.053510] bio: create slab <bio-2> at 2
[ 1388.079385] EXT4-fs (dm-1): mounted filesystem with ordered data mode. Opts: discard
[ 1388.260431] bio: create slab <bio-2> at 2
[ 1388.281332] EXT4-fs (dm-1): mounted filesystem with ordered data mode. Opts: discard
[ 1388.449292] bio: create slab <bio-2> at 2
[ 1388.478583] EXT4-fs (dm-1): mounted filesystem with ordered data mode. Opts: discard
[ 1388.645753] bio: create slab <bio-2> at 2
[ 1388.675497] EXT4-fs (dm-1): mounted filesystem with ordered data mode. Opts: discard
[ 1388.788490] device veth736ca86 entered promiscuous mode
[ 1388.791302] IPv6: ADDRCONF(NETDEV_UP): veth736ca86: link is not ready
[ 1388.827709] IPv6: ADDRCONF(NETDEV_CHANGE): veth736ca86: link becomes ready
[ 1388.829545] docker0: port 1(veth736ca86) entered forwarding state
[ 1388.831030] docker0: port 1(veth736ca86) entered forwarding state
[ 1388.832918] IPv6: ADDRCONF(NETDEV_CHANGE): docker0: link becomes ready
[ 1403.636267] device eth0 entered promiscuous mode
[ 1403.872070] docker0: port 1(veth736ca86) entered forwarding state
[ 1632.212424] docker0: port 1(veth736ca86) entered disabled state
[ 1632.214885] device veth736ca86 left promiscuous mode
[ 1632.216319] docker0: port 1(veth736ca86) entered disabled state
[ 1639.539082] bio: create slab <bio-2> at 2
[ 1643.781465] bio: create slab <bio-2> at 2
[ 1691.024145] device flannel.1 left promiscuous mode
[ 1949.111063] device flannel.1 entered promiscuous mode
[ 1957.101277] bio: create slab <bio-0> at 0
[ 1957.295132] IPv6: ADDRCONF(NETDEV_UP): docker0: link is not ready
[ 1976.120201] bio: create slab <bio-2> at 2
[ 1976.138559] EXT4-fs (dm-1): mounted filesystem with ordered data mode. Opts: discard
[ 1976.296184] bio: create slab <bio-2> at 2
[ 1976.322437] EXT4-fs (dm-1): mounted filesystem with ordered data mode. Opts: discard
[ 1976.360050] bio: create slab <bio-2> at 2
[ 1976.387338] EXT4-fs (dm-1): mounted filesystem with ordered data mode. Opts: discard
[ 1976.461598] device veth603752d entered promiscuous mode
[ 1976.463493] IPv6: ADDRCONF(NETDEV_UP): veth603752d: link is not ready
[ 1976.493306] IPv6: ADDRCONF(NETDEV_CHANGE): veth603752d: link becomes ready
[ 1976.494536] docker0: port 1(veth603752d) entered forwarding state
[ 1976.496077] docker0: port 1(veth603752d) entered forwarding state
[ 1976.497478] IPv6: ADDRCONF(NETDEV_CHANGE): docker0: link becomes ready
[ 1991.520075] docker0: port 1(veth603752d) entered forwarding state
[ 2000.895799] nr_pdflush_threads exported in /proc is scheduled for removal
[ 2000.899488] sysctl: The scan_unevictable_pages sysctl/node-interface has been disabled for lack of a legitimate use case.  If you have one, please send an email to linux-mm@kvack.org.

I don't see any entry for flannel.1 in the output of bridge fdb. That shouldn't be the case right? Please let me know if I'm missing something here.

Thanks!

**Comment by eyakubovich** [comment]
@sanjana-bhat What version of the linux kernel are you running? I'm suspecting that it doesn't have VXLAN DOVE extensions that flannel makes use of (I think you need kernel 3.8.0+). Those extensions generate fdb misses that flannel handles. That would explain why `bridge fdb` doesn't show any entries.

**Comment by sanjana-bhat** [comment]
@eyakubovich, I use kernel version 3.10.0, a RHEL7 host. bridge fdb would show me entries for flannel.1 when I built code from git HEAD. It started crashing soon after I switched to v0.4.1. How do I check if I have the VXLAN DOVE extensions that flannel makes use of?

**Comment by sanjana-bhat** [comment]
I built the code from git HEAD and I now see flannel.1 under bridge fd output. 
a2:59:01:6c:5b:ec dev flannel.1 dst 10.88.217.91 self permanent

With v0.4.1 it seems to be missing.

**Comment by eyakubovich** [comment]
@sanjana-bhat I don't have an easy way to check presence of DOVE extensions but I pulled down CentOS 7 kernel version 3.10.0-123.1.2 source and they are in there. So I'm not sure why the L2 misses are not showing up (try `bridge monitor fdb` to see if it shows them). I also don't know why the kernel is crashing.

Here's some background on v0.4.1 vs HEAD. In v0.4.1 (and prior), flanneld monitors for FDB and ARP misses (it calls them L2 and L3 misses) and dynamically pushes the entries into the kernel (not permanently). The L2 misses are communicated via the DOVE extensions. Since the L2 misses never make it to flanneld in your case, they never get installed and that's why `bridge fdb` shows nothing.

With HEAD, we tried to move to installing permanent entries (both ARP and FDB) when we learn of a new host. This removes the need for monitoring the misses. However, as you discovered, the code does not fully work. We intent to fix it by rolling back to dynamic L3 entries. So FDB entries will be permanent but ARP entries will still be installed on a miss. The good news is that ARP misses are not monitored by DOVE extensions (and they seem to be working for you). We hope this lands in HEAD in a couple of days. The bad news is that there have been a number of big changes since 0.4.1 so it may be a while since we shake out all the bugs and tag 0.5.0.
